src/xenia/base/threading_posix.cc

```diff
@@ -48,265 +97,697 @@ void MaybeYield() {
+    sev.sigev_value.sival_ptr = (void*)&callback_;
+    if (timer_create(CLOCK_REALTIME, &sev, &timer_) == -1) return false;
+
+    // Start timer
+    itimerspec its{};
+    its.it_value = DurationToTimeSpec(period);
+    its.it_interval = its.it_value;
+    return timer_settime(timer_, 0, &its, nullptr) != -1;
   }
 
  private:
   std::function<void()> callback_;
+  timer_t timer_;
 };
 
 std::unique_ptr<HighResolutionTimer> HighResolutionTimer::CreateRepeating(
     std::chrono::milliseconds period, std::function<void()> callback) {
+  install_signal_handler(SignalType::kHighResolutionTimer);
   auto timer = std::make_unique<PosixHighResolutionTimer>(std::move(callback));
   if (!timer->Initialize(period)) {
     return nullptr;
   }
   return std::unique_ptr<HighResolutionTimer>(timer.release());
 }
 
-// TODO(dougvj) There really is no native POSIX handle for a single wait/signal
-// construct pthreads is at a lower level with more handles for such a mechanism
-// This simple wrapper class could function as our handle, but probably needs
-// some more functionality
-class PosixCondition {
+class PosixConditionBase {
  public:
-  PosixCondition() : signal_(false) {
-    pthread_mutex_init(&mutex_, NULL);
-    pthread_cond_init(&cond_, NULL);
+  virtual bool Signal() = 0;
+
+  WaitResult Wait(std::chrono::milliseconds timeout) {
+    bool executed;
+    auto predicate = [this] { return this->signaled(); };
+    auto lock = std::unique_lock<std::mutex>(mutex_);
+    if (predicate()) {
+      executed = true;
+    } else {
+      if (timeout == std::chrono::milliseconds::max()) {
+        cond_.wait(lock, predicate);
+        executed = true;  // Did not time out;
+      } else {
+        executed = cond_.wait_for(lock, timeout, predicate);
+      }
+    }
+    if (executed) {
+      post_execution();
+      return WaitResult::kSuccess;
+    } else {
+      return WaitResult::kTimeout;
+    }
   }
 
-  ~PosixCondition() {
-    pthread_mutex_destroy(&mutex_);
-    pthread_cond_destroy(&cond_);
+  static std::pair<WaitResult, size_t> WaitMultiple(
+      std::vector<PosixConditionBase*> handles, bool wait_all,
+      std::chrono::milliseconds timeout) {
+    using iter_t = decltype(handles)::const_iterator;
+    auto predicate = [](auto h) { return h->signaled(); };
+    auto operation = wait_all ? std::all_of<iter_t, decltype(predicate)>
+                              : std::any_of<iter_t, decltype(predicate)>;
+    auto aggregate = [&handles, operation, predicate] {
+      return operation(handles.cbegin(), handles.cend(), predicate);
+    };
+    bool executed;
+
+    std::unique_lock<std::mutex> lock(PosixConditionBase::mutex_);
+
+    if (aggregate()) {
+      executed = true;
+    } else {
+      if (timeout == std::chrono::milliseconds::max()) {
+        PosixConditionBase::cond_.wait(lock, aggregate);
+        executed = true;
+      } else {
+        executed = PosixConditionBase::cond_.wait_for(lock, timeout, aggregate);
```

Comment:
Might change this to:
```executed = PosixConditionBase::cond_.wait_for(lock, timeout, aggregate) == std::cv_status::no_timeout;```

Reply:
That wouldn't work. `std::condition_variable::wait_for` only returns a `std::cv_status` if there is no predicate. If there is one, the wait returns a `bool`. This case needs a predicate.

Reply:
Ah, you're right - indeed it does.